src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,411 @@
+                throw new SynchronizationLockException();
+            }
+
+            try
+            {
+                Guid invocationId = Guid.NewGuid();
+                bool stillLockedByCurrentThread = await RequestUnlockAsync(sessionId, threadId, invocationId).CfAwait();
+
+                if (stillLockedByCurrentThread)
+                    _lockedThreadToSession.Add(threadId, sessionId);
+                else
+                    _lockedThreadToSession.Remove(threadId);
+
+                _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+            }
+            catch (RemoteException e)
+            {
+                if (e is RemoteException { Error: RemoteError.SessionExpiredException })
+                {
+                    _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                    _lockedThreadToSession.Remove(threadId);
+                }
+                else if (e is RemoteException { Error: RemoteError.IllegalMonitorState })
+                {
+                    _lockedThreadToSession.Remove(threadId);
+                }
+
+                throw;
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+        #endregion
+
+        /// <summary>
+        /// Verifies there is no lock between ThreadId->SessionId. Otherwise throws.
+        /// </summary>
+        /// <param name="threadId"></param>
+        /// <param name="sessionId"></param>
+        /// <param name="releaseSession"></param>
+        /// <exception cref="LockOwnershipLostException"></exception>
+        private void VerifyNoLockOnThread(long threadId, long sessionId, bool releaseSession = true)
+        {
+            if (_lockedThreadToSession.TryGetValue(threadId, out var lockedSessionId) && lockedSessionId != sessionId)
+            {
+                _lockedThreadToSession.Remove(threadId);
+
+                if (releaseSession)
+                    _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+
+                throw new LockOwnershipLostException(lockedSessionId.ToString("D"));
+            }
+        }
+
+        /// <summary>
+        /// Verifies there is no ThreadId -> SessionId map. Otherwise throws.
+        /// </summary>
+        /// <param name="threadId"></param>
+        /// <exception cref="LockOwnershipLostException"></exception>
+        private void VerifyNoLockedSessionExist(long threadId)
+        {
+            if (_lockedThreadToSession.TryGetValue(threadId, out var lockedSessionId))
+            {
+                _lockedThreadToSession.Remove(threadId);
+                throw new LockOwnershipLostException(lockedSessionId.ToString("D"));
+            }
+        }
+
+        public override ValueTask DestroyAsync()
+        {
+            _lockedThreadToSession.Clear();
+            return default;
+        }
+
+        internal class LockOwnershipState
+        {
+            public long Fence { get; private set; }
```

Comment:
You don't need the `private` setters for these properties if the fields are readonly and set by the constructor.

Reply:
Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/590/commits/615acd74e54cd18425eb94dfeb577ca5512e2a61